photon-ml/src/main/scala/com/linkedin/photon/ml/FeatureIndexingJob.scala

```diff
@@ -138,39 +171,159 @@ class FeatureIndexingJob(val sc: SparkContext,
   }
 
   def run(): Unit = {
-    val inputRdd = AvroIOUtils.readFromAvro[GenericRecord](sc, inputPath, 10)
-    val featuresRdd = partitionedUniqueFeatures(inputRdd)
-    buildIndexMap(featuresRdd)
+    val inputRdd = AvroUtils.readAvroFiles(sc, inputPaths, 10)
+
+    ensureOutputPath(outputPath)
+
+    featureShardIdToFeatureSectionKeysMap match {
+      case Some(shardToSectionsMap) =>
+        // If a shard to feature section set was specified, build an index for each shard
+        shardToSectionsMap.map { case (shardId, featureSections) =>
+          val featuresRdd = partitionedUniqueFeatures(inputRdd, Some(featureSections))
+          buildIndexMap(featuresRdd, outputPath, shardId)
+        }
+
+      case _ => {
+        // Otherwise, build a global index
+        val featuresRdd = partitionedUniqueFeatures(inputRdd)
```

Comment:
The more I think about such design choice, the more I feel like that sharded feature maps might not be necessary at all, and a global index map might be suffice, and is much cleaner.

This is because the feature indexing job should be only care about mapping the feature keys from Strings to Ints, and such mapping shouldn't be shard-dependent.

But I am not suggesting that we should make a decision in this PR, just think aloud here.